Small perf improvments.

cijothomas opened the pull request:
1.Caches reflective lookup of method addOnSendingHeadersMethod to improve performance

2. Params for reflective call is a Action, which is now created once , and re-used vs creating new action every time. Again, this is to improve performance, and no functionality change is expected.

Address a portion of https://github.com/Microsoft/ApplicationInsights-dotnet-server/issues/510

SergeyKanzhelev commented on Src/Web/Web.Shared.Net/ApplicationInsightsHttpModule.cs:
```diff
@@ -121,29 +145,15 @@ private void AddCorreleationHeaderOnSendRequestHeaders(HttpApplication httpAppli
-                        };
-
-                        addOnSendingHeadersMethod.Invoke(httpApplication.Response, parameters);
+                        this.addOnSendingHeadersMethod = httpApplication.Response.GetType().GetMethod("AddOnSendingHeaders");
```
Why not in constructor?

cijothomas replied:
will do.

cijothomas replied:
done

SergeyKanzhelev commented on Src/Web/Web.Shared.Net/ApplicationInsightsHttpModule.cs:
```diff
@@ -121,29 +145,15 @@ private void AddCorreleationHeaderOnSendRequestHeaders(HttpApplication httpAppli
+                    
+                    if (this.addOnSendingHeadersMethod != null)
+                    {                        
+                        this.addOnSendingHeadersMethod.Invoke(httpApplication.Response, this.paramsForReflectiveCall);
```
Invoke is more expensive than delegate. See https://gist.github.com/jmangelo/3173109#file-gistfile1-cs-L57-L66 for example use

SergeyKanzhelev commented on Src/Web/Web.Shared.Net/ApplicationInsightsHttpModule.cs:
```diff
@@ -18,6 +18,9 @@ public sealed class ApplicationInsightsHttpModule : IHttpModule
         private readonly ExceptionTrackingTelemetryModule exceptionModule;
+        private MethodInfo addOnSendingHeadersMethod;
+        private bool? addOnSendingHeadersMethodExists;
+        private object[] paramsForReflectiveCall;
```
Name should be reflective of content. This works better: `addOnSendingHeadersMethodParams`

cijothomas replied:
done

cijothomas commented:
@SergeyKanzhelev Addressed the comments. Made invocation using delegate - but only for !Net40 as there was this interface ISubscriptionToken, returned by addOnSendingHeaders(), which is not available in NET40 but is required for constructing the open delegate.

cijothomas commented:
Even without the recent fix (delegate vs invoke), the overhead from 9000 RPS app reduced from 39 to 38. (2 runs made).
Will do more runs with the  recent fix included.
